pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sync"
+	"time"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = bool
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
+
+	pushCounter      uint32
+	debounceCounter  uint32
+	debounceAfterEnd time.Time
+	debounceMaxEnd   time.Time
+	lastEnqueueTime  time.Time
+	lastEmitTime     time.Time
+	debounceAfter    time.Duration
+	debounceMax      time.Duration
+	mutex            sync.Mutex
+
+	merge              func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent
+	accumulator        DebouncedEvent
+	initialAccumulator DebouncedEvent
+
+	running bool
+}
+
+// NewDebouncedChannel create a new instance of DebouncedChannel
+// debounceAfter: A new outgoing event is emitted if no additional event is enqueued during this period (resets on every enqueued event).
+// debounceMax: A new outgoing event is emitted after this period (even if additional events were enqueued in that period).
+// initialAccumulator: Initial value of the accumulator
+// merge: function, which is called to merge two events if they are debounced.
+func NewDebouncedChannel(debounceAfter time.Duration, debounceMax time.Duration,
+	initialAccumulator DebouncedEvent, merge func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent) *DebouncedChannel {
+
+	c := make(chan func(func(DebouncedEvent)), 1)
+
+	return &DebouncedChannel{
+		C:                  c,
+		c:                  c,
+		initialAccumulator: initialAccumulator,
+		accumulator:        initialAccumulator,
+		merge:              merge,
+		debounceAfter:      debounceAfter,
+		debounceMax:        debounceMax,
+		lastEmitTime:       time.Unix(0, 0),
+	}
+
+}
+
+// Enqueue pushes a new event to the channel
+func (dc *DebouncedChannel) Enqueue(element DebouncedEvent) {
+	dc.mutex.Lock()
+	defer dc.mutex.Unlock()
+	dc.accumulator = dc.merge(dc.accumulator, element)
+	if dc.debounceCounter == 0 {
+		dc.debounceMaxEnd = time.Now().Add(dc.debounceMax)
+		go func() {
+			dc.mutex.Lock()
+			defer dc.mutex.Unlock()
```

Comment:
This is not right, once running here, others can not acquire the lock, and the Enqueue will block

Reply:
IMHO it is correct, since we unlock before sleep and lock again after sleep (line 87-89).
Did I miss something?

Also, it works in the tests. Should be fine.

Reply:
wow, i haven't met such writing.

Reply:
I think the confusing thing is we have three levels of nesting all modifying the same lock. After looking at it closely I think it makes sense, but it takes a while to wrap my head around it